File: 2/C++/main.cpp
/**
* @author Matias Grioni
* @date 03:22 17/3/15
* @desc Second Project Euler problem. Finds the sum of all even
*       fibonacci terms lower than some limit.
*/

#include <iostream>

using namespace std;

/**
* Sums even terms of the fibonacci sequence less than the
* parameter passed in.
*
* @param upperLimit - The inclusive bound for the fibonacci terms.
* @return - The sum of even fibonacci terms less than the param.
*/
long sumEvenFib(int upperLimit) {
    // Start the counters assuming the first iteration has been done.
    // The state of these variables would be like this is if
    // a_n = fib(n) at n = 1, and sum is the sum of all even
    // terms, e_k, such that k < n.
    int a = 1;
    int nextA = 2;
    long sum = 2;

    while(nextA < upperLimit) {
        int tmp = nextA;
        nextA += a;
        a = tmp;

        if(nextA % 2 == 0)
            sum += nextA;
    }

    return sum;
}

int main() {
    int upperLimit;
    cout << "Sum even fibonnaci terms less than > ";
    cin >> upperLimit;

    cout << "The sum is " << sumEvenFib(upperLimit);
    return 0;
}

File: 2/Haskell/main.hs
{-
 - Author: Matias Grioni
 - Created: 7/2/2015
 -
 - Solution to Project Euler problem 2 using Haskell. Sums all even valued
 - terms of the Fibonacci sequence less than n, which is input by the user.
 -}

module Main where

-- Fibonacci sequence up to n, assuming n is a natural number.
fibonacci :: Int -> [Int]
fibonacci 1 = [1]
fibonacci 2 = [2, 1]
fibonacci n
    | n < 1 = error "Parameter must be a natural number"
    | otherwise = (head p + (head $ tail p)):p
    where p = fibonacci (n-1)

main = do
    putStrLn "Sums even Fibonacci sequence terms less than n"
    putStrLn "Enter n > "
    input <- getLine
    let n = read input :: Int

    -- Finds the sum of all fibonacci terms less than n that are even
    -- Very inefficient as for each term, the entire fibonacci sequence is
    -- generated and the head of the list is taken. Come back later once
    -- I'm more familiar with the language.
    print $ sum $ filter ((== 0) . (`mod` 2)) $ takeWhile (< n) [head (fibonacci n) | n <- [1..]]
